### PR TITLE
Enforce layering_check in Bazel build rules.

### PR DESCRIPTION
**Pull request by RafaelMarinheiro**
Enforcing layering_check ensures that the Build targets do not rely on transitive dependencies. See
https://github.com/bazelbuild/bazel/pull/11440 for a detailed description of the feature.

We also do a style pass on the build files, ensuring that common linters are happy with it.

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -366,194 +379,334 @@ cc_binary(
...
-cc_library(
-    name = "link_test_common",
-    testonly = 1,
-    srcs = ["test/link/linker_fixture.h"],
-    compatible_with = [],
+# PCH (precompiled header) tests only work when using CMake and MSVC on Windows,
+# so they will be skipped in the Bazel builds.
+
+[cc_test(
+    name = "base_{testcase}_test".format(testcase = f[len("test/"):-len("_test.cpp")]),
```
Comment (dneto0) [inline BUILD.bazel:414]: This is a nice, readable way to express the filename remapping.

**Review by dneto0 — APPROVED** [review]
Thanks!

**Comment by dneto0** [comment]
The bazel builds are failing:

Using bazel-5.0.0
+ path=/bin/bazel-5.0.0
+ echo 'Using bazel-5.0.0'
+ export PATH=/bin/bazel-5.0.0:/bin/clang-10.0.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+ PATH=/bin/bazel-5.0.0:/bin/clang-10.0.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+ return 0
+ unset -f print_glob_matches
+ unset -f export_tool_to_path
+ unset -f use_clang
+ unset -f use_gcc
++ date
+ echo Tue Dec 20 05:05:59 UTC 2022: Build everything...
+ bazel build :all
Tue Dec 20 05:05:59 UTC 2022: Build everything...
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
Loading:
Loading: 0 packages loaded
ERROR: Traceback (most recent call last):
	File "/tmpfs/src/github/SPIRV-Tools/BUILD.bazel", line 41, column 21, in <toplevel>
		generate_core_tables(version = "unified1")
	File "/tmpfs/src/github/SPIRV-Tools/build_defs.bzl", line 78, column 26, in generate_core_tables
		).format(**(grammars | outs))
Error: unsupported binary operation: dict | dict


Does this change require bazel after 5.0.0?  That's what our bot VM has in it.

File: BUILD.bazel (hunk 1 above)
Comment (RafaelMarinheiro) [reply BUILD.bazel:414]: Thanks!

**Comment by RafaelMarinheiro** [comment]
`dict | dict` does not work in Bazel 5.0.0. I added a .bazelversion file to document the version of Bazel to work (and also to make tools like Bazelisk choose the right version) and then I fixed build_defs.bzl.

**Review by dneto0 — APPROVED** [review]
Looks good now.  Thanks!

Seems bazel 6.0.0 was released yesterday.
But our VM is not yet updated, so these changes are good.

**Pull request merged**